File: chapters/basics/test_client.rb
$LOAD_PATH.unshift File.expand_path("../", __FILE__)
require_relative './proto/sfapi_pb'
require_relative './proto/sfapi_services_pb'

stub = Starfriends::Stub.new("localhost:8080", :this_channel_is_insecure)

# Now we can use the stub to make RPCs
request = GetFilmRequest.new(id: '4')
begin
  response = stub.get_film(request,
      metadata: {'who' => 'starfriends-ruby-client', 'version' => ['v1','v2']})
  puts response.inspect
rescue GRPC::BadStatus => e
  # RPC exceptions have a numeric 'code' field (the gRPC error code)
  # and a string 'details' (the message supplied by the server)
  puts "failed: #{e.details} (code = #{e.code})"
end

# We'll make another request and also print the response metadata
request = GetFilmRequest.new(id: '7')
op = stub.get_film(request, return_op: true)
begin
  op.execute
  puts response.inspect
rescue GRPC::BadStatus => e
  # RPC exceptions have a numeric 'code' field (the gRPC error code)
  # and a string 'details' (the message supplied by the server)
  puts "failed: #{e.details} (code = #{e.code})"
end
puts "server provided headers: #{op.metadata}"
puts "server provided trailers: #{op.trailing_metadata}"

File: chapters/basics/test_client.go
package main

import (
  "fmt"
  "os"

  "golang.org/x/net/context"
  "google.golang.org/grpc"
  "google.golang.org/grpc/metadata"

  "./proto"
)

func main() {
  conn, err := grpc.Dial("localhost:8080", grpc.WithInsecure())
  if err != nil {
    panic(err)
  }
  stub := proto.NewStarfriendsClient(conn)

  // Now we can use the stub to make RPCs
  ctx := metadata.NewOutgoingContext(context.Background(),
      metadata.Pairs("who", "starfiends-go-client", "version", "v1"))
  req := &proto.GetFilmRequest{Id: "4"}
  resp, err := stub.GetFilm(ctx, req)
  if err != nil {
    fmt.Fprintf(os.Stderr, "RPC failed: %v\n", err)
  } else {
    fmt.Println(resp)
  }

  // We'll make another request and also print the response metadata
  req = &proto.GetFilmRequest{Id: "7"}
  var respHdrs, respTrlrs metadata.MD
  resp, err = stub.GetFilm(ctx, req,
      grpc.Header(&respHdrs), grpc.Trailer(&respTrlrs))
  if err != nil {
    fmt.Fprintf(os.Stderr, "RPC failed: %v\n", err)
  } else {
    fmt.Println(resp)
  }
  fmt.Printf("Server sent headers: %v\n", respHdrs)
  fmt.Printf("Server sent trailers: %v\n", respTrlrs)
}
